---
title: "for_GWAS"
author: "Ruijuan Li"
date: "3/20/2017"
output: html_document
---

# phoenotype data 
```{r}
phenotype.data <- read.csv("~/Desktop/GWAS_manuscript/GWAS/GWAS_analysis/phenotype_data.csv", header = T)

head(phenotype.data)
phenotype.data.2 <- phenotype.data[phenotype.data$FI<150,]

hist(phenotype.data.2$FI, breaks = 15, col = "grey", ylim = c(0, 70))

png("~/Desktop/GWAS_manuscript/GWAS/GWAS_analysis/output/figure/phenotype.png", width=8, height=6, units="in", res=300)
hist(phenotype.data.2$FI, breaks = 15, col = "grey", ylim = c(0, 70), xlab = "Female index", ylab="Number of accessions", main="")
dev.off() 
```

# manhattan plot
```{r}
associattion.result <- read.csv("~/Desktop/GWAS_manuscript/BMC_genetics/Additional_file_4.csv", header = T)

#install.packages("qqman")
library(qqman)

head(associattion.result)
# use p.adjust() to adjust p-value here, this doesn't work 
sum(p.adjust(associattion.result$P, "none") < 0.0001)

## check R squared values 
associattion.result[associattion.result$P < 0.001,] %>%
  mutate(diff = Rsquare.of.Model.with.SNP-Rsquare.of.Model.without.SNP)
```

#### addtive effect
find the resistance allele, and test for additive effect, r square value. 
```{r} 
library(tidyverse)
# determine which is the resistant allele, alleles that decrease trait value 
# import allele estimate 
allele_estimate <- read.csv("~/Desktop/working_manuscript/GWAS_manuscript/BMC_genetics/GWAS_permutation/GAPIT..FI.Allelic_Effect_Estimates.csv", header = T)
head(allele_estimate) 

# GWAS result 
associattion.result <- read.csv("~/Desktop/working_manuscript/GWAS_manuscript/BMC_genetics/Additional_file_4.csv", header = T)
sig_SNP <- associattion.result[which(associattion.result$P < 0.001),]$SNP

# get allele effect for sig SNPs
allele_estimate[which(allele_estimate$SNP %in% sig_SNP),] # S1_352262064 having monor allele as resistance allele 

# compare genotype data with major & minor allele for the 15 loci to look for relationship 
myY=read.table("~/Desktop/working_manuscript/GWAS_manuscript/GWAS/GWAS_analysis/data/association/GAPIT/myGAPIT/246_FI.txt", head=TRUE)
head(myY)

myG=read.table("~/Desktop/working_manuscript/GWAS_manuscript/GWAS/GWAS_analysis/data/association/GAPIT/myGAPIT/246_7220_chr14_sorted.hmp.txt", head=T)
myG[1:10, 1:10]

sig_SNP_G <- 
myG[which(myG$rs %in% sig_SNP),]

sig_SNP_G.2 <- 
sig_SNP_G %>%
  separate(alleles, into=c("major", "minor")) 

## site summary including major & minor allele info 
site_summary <- read.delim("~/Desktop/working_manuscript/GWAS_manuscript/GWAS/GWAS_analysis/data/association/GAPIT/myGAPIT/site_summary_arboreum.txt", header = T)
head(site_summary)
site_summary[which(site_summary$Site.Name %in% sig_SNP),][,c("Site.Name","Major.Allele", "Minor.Allele")]

## now sure in hapmap file, alleles column, before / is major allele, after / is minor allele. 

sig_SNP_G.2
myY
# need to combine these two information  
head(sig_SNP_G.2)
sig_SNP_G.2$resistance <- sig_SNP_G.2$major
sig_SNP_G.2[sig_SNP_G.2$rs=="S1_352262064",]$resistance <- sig_SNP_G.2[sig_SNP_G.2$rs=="S1_352262064",]$minor

# replace allele symbol to resistance allele symbol  
sig_SNP_G.2[1:10, 1:15]

for(i in 1:nrow(sig_SNP_G.2)){{
  for(j in grep("PI", colnames(sig_SNP_G.2))){
    sig_SNP_G.2[i,j] <- ifelse(sig_SNP_G.2[i,j]==sig_SNP_G.2$resistance[i], "R", "N")
  }
}
}

dim(sig_SNP_G.2) 
sig_SNP_G.2[1:10, 1:10]

# combine resistance allele info with phenotype data 
sig_SNP_G.2.t.df <- data.frame(t(sig_SNP_G.2[,grep("PI", colnames(sig_SNP_G.2), value = T)]))
Taxa <- rownames(sig_SNP_G.2.t.df)
sig_SNP_G.2.t.df <- data.frame(sapply(sig_SNP_G.2.t.df, function(x) sub("R", "1", x)))
sig_SNP_G.2.t.df <- data.frame(sapply(sig_SNP_G.2.t.df, function(x) sub("N", "0", x)))
sig_SNP_G.2.t.df <- data.frame(sapply(sig_SNP_G.2.t.df, function(x) as.numeric(as.character(x))))

rownames(sig_SNP_G.2.t.df) <- Taxa
sig_SNP_G.2.t.df

sig_SNP_G.2.t.df.2 <- 
as.data.frame(rowSums(sig_SNP_G.2.t.df))

sig_SNP_G.2.t.df.2$Taxa <- rownames(sig_SNP_G.2.t.df.2) 
colnames(myY) <- c("Taxa", "FI")

sig_SNP_G.2.t.df.3 <- 
sig_SNP_G.2.t.df.2 %>%
  left_join(myY, by="Taxa") 

colnames(sig_SNP_G.2.t.df.3)[1] <- "R.number" 

additive_estimate <- 
sig_SNP_G.2.t.df.3 %>%
  group_by(R.number) %>%
  summarise(mean(FI)) %>%
  as.data.frame()

plot(additive_estimate$`mean(FI)`~additive_estimate$R.number)
additive_estimate
  
boxplot(sig_SNP_G.2.t.df.3$FI~sig_SNP_G.2.t.df.3$R.number, xlab="number of resistance alleles", ylab="female index", col="grey") 

## fit a linear model, regression 
plot(new$all_FI~new$all_number, xlab="number of resistance alleles", ylab="female index", col="grey") 
plot(sig_SNP_G.2.t.df.3$FI~sig_SNP_G.2.t.df.3$R.number, xlab="number of resistance alleles", ylab="female index", col="grey") 
abline(a = 78, b = -3)

fit=lm(sig_SNP_G.2.t.df.3$FI ~ sig_SNP_G.2.t.df.3$R.number)
summary(fit)
confint(fit)
fit$coefficients 

## check residual plot 
res <- resid(fit)
plot(sig_SNP_G.2.t.df.3$R.number, res)
abline(0, 0)
hist(res) # almost normally distributed 

### how to analyze this dataset? 
cor.test(sig_SNP_G.2.t.df.3$R.number, sig_SNP_G.2.t.df.3$FI) 

### how many accessions have fewer than 6 resistance alleles
sum(sig_SNP_G.2.t.df.3$R.number < 6)
  
## additive effect test 
new <- read.csv("~/Desktop/working_manuscript/GWAS_manuscript/GWAS/GWAS_analysis/data/allele_effect/additive_effect/FI_resistance_allele_number.csv", sep = ",")
View(new) 
boxplot(new$all_FI~new$all_number, xlab="number of resistance alleles", ylab="female index", col="grey") 

new[which(new$all_number=="3"),] # 2 
new[which(new$all_number=="5"),] # 2
new[which(new$all_number=="4"),] # 3 
new[which(new$all_number=="2"),] # 4
new[which(new$all_number=="6"),] # 4   
```

# import data 
```{r}
gene.list.arboreum <- read.csv("~/Desktop/GWAS/GWAS_analysis/GO_enrichment/gene_list.csv", header = T)
str(gene.list.arboreum)
colnames(gene.list.arboreum)[1] <- "GeneID"
head(gene.list.arboreum)

# Download the GO annotation and gene length for G.arboreum on terminal end
# Format data for GOseq
library(goseq)
go.terms.arboreum <- read.delim("~/Desktop/GWAS/GWAS_analysis/GO_enrichment/G.arboreum_BGI-CGP_v1.0_genes2GO_modified.txt",header=FALSE,as.is=TRUE)
head(go.terms.arboreum, 100)
tail(go.terms.arboreum)

# reformat GO terms, keep the duplicated gene IDs 
go.terms.arboreum$V1 <- gsub("_BGI-A2_v1.0", "", go.terms.arboreum$V1)
go.terms.arboreum$V4 <- gsub(":", "", go.terms.arboreum$V4)
go.terms.arboreum$V4 <- gsub("GO", "GO:", go.terms.arboreum$V4)
go.terms.arboreum <- go.terms.arboreum[,c("V1", "V4")]
names(go.terms.arboreum) <- c("GeneID","GO")
go.terms.arboreum$GO <- gsub("|", ",", go.terms.arboreum$GO, fixed = T) 
head(go.terms.arboreum)
class(go.terms.arboreum$GO)

# what if combine all duplciated gene IDs (how)
go.terms.arboreum <-aggregate(go.terms.arboreum$GO,list(go.terms.arboreum$GeneID), paste, collapse=",")
head(go.terms.arboreum)
tail(go.terms.arboreum)
names(go.terms.arboreum) <- c("GeneID", "GO")

# all arboreum genes 
all.genes.arboreum <- read.delim("~/Desktop/GWAS/GWAS_analysis/GO_enrichment/arboreum_gene_ID.txt",as.is=TRUE)
head(all.genes.arboreum)
tail(all.genes.arboreum)
names(all.genes.arboreum) <- "GeneID"

gene.lengths.arboreum <- read.delim("~/Desktop/GWAS/GWAS_analysis/GO_enrichment/gene_ID_start_end",as.is=TRUE, header = F)
head(gene.lengths.arboreum)
gene.lengths.arboreum$length <- gene.lengths.arboreum$V5- gene.lengths.arboreum$V4 + 1
gene.lengths.arboreum <- gene.lengths.arboreum[,c(1,7)]
names(gene.lengths.arboreum)[1] <- "geneID"
summary(gene.lengths.arboreum) 

#we need to reduce the gene.length to only contain entries for those genes in our expressed.genes set.  We also need this as a vector
gene.lengths.vector.arboreum <- gene.lengths.arboreum$length[gene.lengths.arboreum$geneID %in% all.genes.arboreum$GeneID]
names(gene.lengths.vector.arboreum) <- gene.lengths.arboreum$geneID[gene.lengths.arboreum$geneID %in% all.genes.arboreum$GeneID]
head(gene.lengths.vector.arboreum)
tail(gene.lengths.vector.arboreum)
length(gene.lengths.vector.arboreum)

# Format go.terms for goseq. We want them in list format, and we need to separate the terms into separate elements.
go.list.arboreum <- strsplit(go.terms.arboreum$GO,split=",")
names(go.list.arboreum) <- go.terms.arboreum$GeneID
head(go.list.arboreum, 100)
tail(go.list.arboreum)
length(go.list.arboreum)

# Format gene expression data for goseq. 
# We need a vector for each gene with 1 indicating differential expression and 0 indicating no differential expression.
rownames(gene.list.arboreum) <- gene.list.arboreum$GeneID
rownames(gene.list.arboreum)
list.arboreum <- all.genes.arboreum$GeneID %in% rownames(gene.list.arboreum) 
sum(list.arboreum)
#for each gene in expressed gene, return FALSE if it is not in DEgene.trt and TRUE if it is.
names(list.arboreum) <- gene.match
head(list.arboreum)
list.arboreum <- as.numeric(list.arboreum) #convert to 0s and 1s
head(list.arboreum)
sum(list.arboreum) # number of genes 

# Calculate over-representation
#determines if there is bias due to gene length.  The plot shows the relationship.
nullp.result.arboreum <- nullp(DEgenes = list.arboreum,bias.data = gene.lengths.vector.arboreum)

#calculate p-values for each GO term
rownames(nullp.result.arboreum) <- names(gene.lengths.vector.arboreum) #because of a bug in nullp()
GO.out.arboreum <- goseq(pwf = nullp.result.arboreum,gene2cat = go.list.arboreum,test.cats=("GO:BP"), use_genes_without_cat=T)
?goseq

#list over-represented GO terms (p < 0.05)
tmp <- GO.out.arboreum[GO.out.arboreum$over_represented_pvalue < 0.05,]
enriched_GO_term <- tmp[grep("BP", tmp$ontology),]
enriched_GO_term
write.csv(enriched_GO_term, file = "~/Desktop/GWAS/GWAS_analysis/GO_enrichment/enriched_GO_term.csv")

# GO visualization
print(GO.out.arboreum[GO.out.arboreum$over_represented_pvalue < 0.05,1:2],row.names=FALSE)

```

# GWAS, redo with permutation  
```{r}
library(multtest)
library(gplots)
# library(LDheatmap) # this could not be succesfully installed 
library(genetics)
library(EMMREML)
library(compiler) #this library is already installed in R
library("scatterplot3d")

# install GAPIT package 
source("http://zzlab.net/GAPIT/previous/gapit_functions_20151103.txt")
# intall EMMA package 
source("http://zzlab.net/GAPIT/emma.txt")

# set working directory 
setwd("~/Desktop/GWAS_manuscript/BMC_genetics/GWAS_permutation/")

myY=read.table("~/Desktop/GWAS_manuscript/GWAS/GWAS_analysis/data/association/GAPIT/myGAPIT/246_FI.txt", head=TRUE)
myG=read.table("~/Desktop/GWAS_manuscript/GWAS/GWAS_analysis/data/association/GAPIT/myGAPIT/246_7220_chr14_sorted.hmp.txt", head=F)
myGAPIT <- GAPIT(Y=myY, G=myG, Model.selection=TRUE, PCA.total=0, Major.allele.zero = T, kinship.algorithm=c("EMMA"))

### start permutation 
# use permutation to get a reasonable threshold. 
# 1) permuate the phenotypic data, equal to doing GWAS on 1000 traits 
# 2) order P-value for all 1000 traits (including the ture trait)
# 3) get the 95% quantile, set that as the threshold for p-value  
```

### only look at the genes in the 72kb flanking region instead of 300kb 
```{r}


```

# session info
```{r}
sessionInfo() 
```  

